src/accounts.rs

```diff
@@ -1,216 +1,336 @@
-    /// The status of a ledger scan for this account's key.
-    ///
-    /// If a ledger scan using this account's key is in progress, `scan_status` contains the index
-    /// of the next event to be scanned and the index of the last event in the scan's range of
-    /// interest, in that order. Note that the former may be greater than the latter, since the scan
-    /// will not complete until it has caught with the main event loop, which may have advanced past
-    /// the end of the range of interest.
-    pub scan_status: Option<(EventIndex, EventIndex)>,
+/// Accounts stored in an accounts store.
+pub struct Accounts<L: Ledger, Key: KeyPair + DeserializeOwned + Serialize> {
+    /// A key-value store for accounts.
+    store: AccountsStore<L, Key, Key::PubKey>,
 }
 
-impl<Key: KeyPair> AccountInfo<Key> {
-    pub fn new<L: Ledger>(
-        account: Account<L, Key>,
-        assets: Vec<Asset>,
-        records: Vec<RecordInfo>,
-    ) -> Self {
-        let mut balances = HashMap::new();
-        for rec in &records {
-            *balances
-                .entry(rec.ro.asset_def.code)
-                .or_insert_with(U256::zero) += rec.amount().into();
-        }
-        Self {
-            address: account.key.pub_key(),
-            description: account.description,
-            used: account.used,
-            balances,
-            assets,
-            records,
-            scan_status: account.scan.map(|scan| (scan.status())),
-        }
+impl<L: Ledger, Key: KeyPair + DeserializeOwned + Serialize> Accounts<L, Key> {
+    #![allow(dead_code)]
+
+    /// Load an accounts store.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        loader: &mut AtomicStoreLoader,
+        adaptor: EncryptingResourceAdapter<(Key::PubKey, Option<Account<L, Key>>)>,
+        pattern: &str,
+        fill_size: u64,
+    ) -> Result<Self, KeystoreError<L>> {
+        let log = AppendLog::load(
+            loader,
+            adaptor,
+            &format!("keystore_{}_accounts", pattern),
+            fill_size,
+        )?;
+        let store = AccountsStore::<L, Key, Key::PubKey>::new(log)?;
+        Ok(Self { store })
     }
-}
 
-impl<Key: KeyPair> PartialEq<Self> for AccountInfo<Key> {
-    fn eq(&self, other: &Self) -> bool {
-        self.address == other.address
-            && self.description == other.description
-            && self.used == other.used
-            && self.assets == other.assets
-            && self.records == other.records
-            && self.balances == other.balances
-            && self.scan_status == other.scan_status
+    /// Iterate through the accounts.
+    pub fn iter(&self) -> impl Iterator<Item = Account<L, Key>> + '_ {
+        self.store.iter().cloned()
     }
-}
 
-/// Keys with a public part.
-pub trait KeyPair: Clone + Send + Sync {
-    type PubKey: std::hash::Hash + Eq;
-    fn pub_key(&self) -> Self::PubKey;
-}
+    /// Get the keys of all accounts.
+    pub fn keys(&self) -> Vec<Key> {
```

Comment:
How about `-> impl Iterator<Item = &Key>` to avoid `clone` and `collect` (the caller can always do `keys().cloned().collect()` if they want)

Reply:
Actually might not be possible to do `&Key` because it would borrow from a temporary value, but we can at least avoid an unnecessary `collect`